### ISSUE TITLE
Feature Request: Add linting to `template/contracts/`

### ISSUE THREAD

**Issue by kyhyco** [coinbase/build-onchain-apps#443]
### Describe the solution you'd like

Took a look at linters for solidity:

1. [solhint](https://github.com/protofire/solhint)
`solhint` is a linter with [rules](https://github.com/protofire/solhint/blob/develop/docs/rules.md). Similar control mechanisms to `eslint`, supporting disabling rules via comments and allow you to create new rules.

`solhint` provides additional styling rules that augments options mentioned in `Formatters`. Pretty neat.

2. [ethlint](https://github.com/duaraghav8/Ethlint)
Project appears to be inactive

-------------------------------------

`ethlint` is out as it is inactive. 

For `solhint`, we need to add two files: `.solhint.json` and `.solhintignore`.

In `.solhint.json`, we need to decide which rules sets to follow. `solhint:recommended` option looks good but would like to hear from more experienced solidity developers on which rule sets to enable / how their experience with `solhint` has been.

In `.solhintignore`, `lib` should be added as we don't want to lint third party imports.

Been using `solhint` for a couple days and it's been decent so far.

**Comment by awilliams1-cb** [comment]
Yeah besides `forge fmt`, solhint is the only other linter I've seen used before. `forge fmt` does have its own [section](https://book.getfoundry.sh/reference/config/formatter) in the config file that we could also set to get more use out of it. Regarding which rules to use for solhint, the recommended [Best Practices](https://protofire.github.io/solhint/docs/rules.html#best-practise-rules) and [Style Guide Rules](https://protofire.github.io/solhint/docs/rules.html#style-guide-rules) sections look good. However, the [Security Rules](https://protofire.github.io/solhint/docs/rules.html#security-rules) look pretty out of date / not of high fidelity to be worth enabling.

**Comment by danimesq** [comment]
Additional suggestion for tests: Use Branching Tree Technique for Solidity tests, through bulloak (https://github.com/coinbase/build-onchain-apps/issues/523)
